Enhance the calculation of link quality metrics (#175)

abtink opened the pull request:
This commit adds a new class `LinkQualityInfo` which stores all relevant information about quality of a link, including average received signal strength (RSS), link margin, and link quality value. The RSS average is obtained using an adaptive exponential moving average filter.

The Neighbor class now includes an instance of LinkQualityInfo. The RSS average is updated from `mac.cpp` whenever a valid/verified packet is received from the corresponding neighbor. The  mesh_forwarder, mle_router, and mle modules are modified to use the new methods from the newly added class. 

This change addresses issue #175.

pvanhorn commented on src/core/thread/topology.hpp:
```diff
@@ -81,7 +82,7 @@ class Neighbor
     bool    mPreviousKey : 1;            ///< Indicates whether or not the neighbor is still using a previous key
     bool    mDataRequest : 1;            ///< Indicates whether or not a Data Poll was received
-    int8_t  mRssi;                       ///< Received Signal Strength Indicator
+    LinkQualityInfo mLinkInfo;           ///< Link quality info (contains average RSS, link margin and link quality).
```
variable name alignment.

jwhui commented on src/core/thread/link_quality.cpp:
```diff
@@ -0,0 +1,274 @@
+    uint16_t    newValue;
+    uint16_t    oldAverage;
+
+    VerifyOrExit(anRss < 0, error = kThreadError_InvalidArgs);
```
Are non-negative RSS values really invalid?

abtink replied:
Good question. 

Practically I don't expect any 15.4 radio to give us a positive RSS. The main reason I restricted the positives was that I wanted to the average to fit in 10 bits (with the goal to keep the total link info data per neighbor to fit in 16 bit).

How about if we see a positive RSS we max/cut it to zero (and not consider it invalid)? Does this make sense?

jwhui replied:
I think it's safer to set a rail, rather than consider non-negative values invalid.

jwhui commented on src/core/thread/link_quality.cpp:
```diff
@@ -0,0 +1,274 @@
+    newValue = -anRss;
+    newValue <<= 3;
+
+    if (mCount < 8)
```
Magic number.

jwhui commented on src/core/thread/link_quality.cpp:
```diff
@@ -0,0 +1,274 @@
+    {
+        VerifyOrExit(aBufferLen >= sizeof(kUnknownRssString), error = kThreadError_NoBufs);
+
+        strcpy(aCharBuffer, kUnknownRssString);
```
strncpy?

abtink replied:
We check the buffer len to see if we have space before `strcpy` but I agree that `strncpy` will still be safer. Will change it

jwhui commented on src/core/thread/link_quality.cpp:
```diff
@@ -0,0 +1,274 @@
+    "000", "125", "250", "375", "500", "625", "750", "875"
+};
+
+LinkQualityInfo LinkQualityInfo::sNoiseFloorAverage;
```
Better not to make `sNoiseFloorAverage` static if we want to support multiple radios in the future.

jwhui commented on src/core/thread/link_quality.cpp:
```diff
@@ -0,0 +1,274 @@
+
+    // Apply the hysteresis threshold based on the last link quality value.
+
+    switch (aLastLinkQuality)
```
Couple comments
1. The spec is being updated to have the hysteresis centered on the target threshold.
2. The hysteresis should be applied regardless of what `aLastLinkQuality` was.

For moving up to 3: >= 21 dB
For moving down to 2: <= 19 dB
For moving up to 2: >= 11 dB
For moving down to 1: <= 9 dB
For moving up to 1: >= 3 dB
For moving down to 0: <= 1 dB

abtink replied:
Nice! Makes more sense to apply the hysteresis on both sides (up and down). When I read the spec, this was what I initially expected +1/-1 dB around the threshold, but the example used +12 dB for move to 2, so I just applied the hysteresis on the upside. I guess we still need some memory (last value) or  to determine if we are moving up or down, right? 

I think I keep the current implementation for now... Should be easy to change it when the spec is updated. Makes sense?

jwhui replied:
I think it's fine to keep the thresholds uncentered for now.  However, I think the intent of the hysteresis is to apply the +2 dB whenever moving to a higher link quality.  The logic below seems to only apply the hysteresis to the next threshold above.

abtink replied:
I see. Got it... will change it.

abtink replied:
Changed this in the latest commit (2aa489e).

jwhui commented on src/core/thread/link_quality.hpp:
```diff
@@ -0,0 +1,241 @@
+ * exponential moving average filter.
+ *
+  */
+OT_TOOL_PACKED_BEGIN
```
We currently only used `OT_TOOL_PACKED_*` for structures used in protocol messages.

jwhui commented on src/core/thread/mle_router.cpp:
```diff
@@ -860,9 +866,10 @@ ThreadError MleRouter::HandleLinkAccept(const Message &aMessage, const Ip6::Mess
-                                                                (aMessageInfo.mLinkInfo)->mLinkMargin);
+        mRouters[routerId].mLinkQualityOut =
+            LinkQualityInfo::ConvertLinkMarginToLinkQuality(linkMargin.GetLinkMargin());
+        mRouters[routerId].mLinkQualityIn =
```
Does `mLinkInfo.GetLinkMargin()` take the place of `mLinkQualityIn`?  Better not to maintain two separate values.

abtink replied:
Good point. I think we can replace/remove `mLinkQualityIn`. Let me check where/how it's used to make sure.

jwhui commented:
Thanks for working on this - this is great.  See my inline comments.

Also, I know OpenThread is light on unit testing in general, but @pvanhorn started a new precedent.  Can we add some unit tests for `LinkQualityInfo`?

abtink commented:
In the new commit c6eb92d addressed the PR suggestions.
Also added a unit-test to test/verify the newly added `LinkQualityInfo` class (the test check the class API, averaging algorithm, link quality calculation, ...).

jwhui commented:
LGTM 👍
